File: partitionString.py
class Solution:

# Approach 1- Greedy method
    def partitionString(self, s): # TC O(n) // SC O(1)
        n=len(s)
        lastSeen=[-1]*26
        count=1
        subStringStart=0
        for i in range(n):
            if lastSeen[ord(s[i])-ord('a')]>=subStringStart:
                count+=1
                subStringStart=i
            
            lastSeen[ord(s[i])-ord('a')]=i 
        
        return count

    def partitionString1(self, s):
        seen=set()
        count=1
        for c in s:
            if c in seen:
                seen=set()
                count+=1
            seen.add(c)
        return count

s=Solution()
print(s.partitionString("abacaba"))
print(s.partitionString1("abacaba"))

File: partitionString.java
import java.util.*;
// Approach 1- Greedy method
class Solution{ // TC O(n) SC O(1)
	public int partitionString(String s){
		int[] lastSeen=new int[26];
		int count=1, lastSubstring=0;
		Arrays.fill(lastSeen, -1);
		for(int i=0; i<s.length(); i++){
			if(lastSeen[s.charAt(i)-'a']>=lastSubstring){
				count++;
				lastSubstring=i;
			}
			lastSeen[s.charAt(i)-'a']=i;
		}
		return count;
	}
}

class partitionString{
	public static void main(String[] args) {
		String x="abacaba";
		Solution s= new Solution();
		System.out.println(s.partitionString(x));
	}
}